Привет, меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>


<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
- Отлично, что стоп-слова были исключены при векторизации!
- Я оставил некоторые советы, обрати на них внимание. Надеюсь они будут полезными или интересными
- Ты успешно справилась с задачей машинного обучения для текстов, поздравляю!
* Отправляю проект назад, чтобы ты дополнила работу(по желанию) и задала вопросы, если они у тебя есть. Если их нет, то можешь просто отправить проект еще раз и я его зачту.



<font color='green'><b>Полезные (и просто интересные) материалы:</b> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Подготовка-текста" data-toc-modified-id="Подготовка-текста-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подготовка текста</a></span></li><li><span><a href="#Разбиение-на-выборки" data-toc-modified-id="Разбиение-на-выборки-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Разбиение на выборки</a></span></li><li><span><a href="#Векторизация" data-toc-modified-id="Векторизация-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Векторизация</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Dummy-Classifier" data-toc-modified-id="Dummy-Classifier-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Dummy Classifier</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#CatBoostClassifier" data-toc-modified-id="CatBoostClassifier-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>CatBoostClassifier</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

<div class="alert alert-block alert-success">
<b>Успех:</b>  спасибо за описание проекта
</div>

## Подготовка

In [91]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.ensemble import RandomForestClassifier
from catboost import CatBoostClassifier
from sklearn.dummy import DummyClassifier
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score 
from pymystem3 import Mystem
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.utils import shuffle
import time

import re
import nltk
from nltk.corpus import stopwords as nltk_stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer 

nltk.download('wordnet')
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в верхней части ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [69]:
df = pd.read_csv('/datasets/toxic_comments.csv')
df.head(3)

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
text     159571 non-null object
toxic    159571 non-null int64
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [71]:
df.describe()

toxic
count  159571.000000
mean        0.101679
std         0.302226
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000

Загруженные данные соответсвуют заданию. Пропущенных значений нет. Доля токсичных комментариев всего 0,1. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены и осмотрены
</div>
<div class="alert alert-block alert-info">
<b>Совет:</b> В задачах классификации, всегда желательно проверять дисбаланс классов

### Подготовка текста

На примере первых комментариев, можно видеть, что в комментариях встречаются знаки припенания и заглавные буквы, готовя текст к анализу лучше избавимся от этого и лемматизируем его. 

In [72]:
m = Mystem()
def lemmatize_(text):
    text = text.lower()
    lemms = "".join(m.lemmatize(text))
    clear_text = re.sub(r'[^a-zA-Z]', ' ', lemms) 
    return " ".join(clear_text.split())

In [73]:
df['lemm_text'] = df['text'].apply(lemmatize_)
df.head(3)

text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   
2  Hey man, I'm really not trying to edit war. It...      0   

                                           lemm_text  
0  explanation why the edits made under my userna...  
1  d aww he matches this background colour i m se...  
2  hey man i m really not trying to edit war it s...

<div class="alert alert-block alert-info">
<b>Совет: </b> Очистку желательно было провести до лемматизации. Иначе, слова с точкой не лемматизируются.
    </div>
<div class="alert alert-block alert-success">
<b>Успех: </b> Приятно видеть результаты до/после


### Разбиение на выборки

Сделаем стратификацию выборок по признаку токсичности комментариев, так как у нас есть сильный дисбаланс классов.

In [74]:
features = df['lemm_text']
target = df['toxic']

features_train, features_test, target_train, target_test = train_test_split(features, target, stratify= target, test_size=0.4, random_state=12345)

### Векторизация

Вычислим TF-IDF для нашего корпуса текстов. Укажиv стоп-слова. Векторизатор будем обучать только на тренировочной выборке, чтобы модель не знала заранее частотность слов в тестовой выборке.

In [75]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [76]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords) 

In [77]:
features_train = count_tf_idf.fit_transform(features_train.values.astype('U'))

features_test = count_tf_idf.transform(features_test.values.astype('U'))

In [78]:
print(features_train.shape)
print(features_test.shape)

(95742, 125904)
(63829, 125904)


<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что обучила векторизатор только на тренировочной чатси данных. Это уменьшает переобучение.
</div>

## Обучение

### Dummy Classifier

In [79]:
dummy_clf = DummyClassifier(random_state=12345, strategy="stratified")

dummy_clf.fit(features_train, target_train)

print('F1 Dummy Classifier на тренировочной выборке', f1_score(target_train, dummy_clf.predict(features_train)))

F1 Dummy Classifier на тренировочной выборке 0.10643464184540671


In [80]:
%%time
start_time = time.time()
print('F1 Dummy Classifier на тестовой выборке', f1_score(target_test, dummy_clf.predict(features_test)))

time_dm = (time.time() - start_time)
time_dm

F1 Dummy Classifier на тестовой выборке 0.10680644181099971
CPU times: user 40.7 ms, sys: 1.51 ms, total: 42.2 ms
Wall time: 49.6 ms


0.0494692325592041

### Логистическая регрессия

In [81]:
%%time
start_time = time.time()

model_lr = LogisticRegression(class_weight='balanced')
model_lr.fit(features_train, target_train)

print('F1 логистической регрессии на тренировочной выборке', f1_score(target_train, model_lr.predict(features_train)))

time_lr = (time.time() - start_time)
time_lr

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


F1 логистической регрессии на тренировочной выборке 0.8439373601789709
CPU times: user 6.85 s, sys: 4.85 s, total: 11.7 s
Wall time: 11.8 s


11.796871185302734

In [82]:
train_f1 = cross_val_score(model_lr, 
                      features_train, 
                      target_train, 
                      cv=3, 
                      scoring='f1').mean()

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [83]:
%%time
start_time = time.time()
print('F1 логистической регрессии на тестовой выборке', f1_score(target_test, model_lr.predict(features_test)))

time_lr_pred = (time.time() - start_time)
time_lr_pred

F1 логистической регрессии на тестовой выборке 0.7513606613847743
CPU times: user 31.5 ms, sys: 533 µs, total: 32 ms
Wall time: 46 ms


0.04588580131530762

<div class="alert alert-block alert-info">
<b>Совет: </b> Тут можно было подобрать параметр C.
</div>

### DecisionTreeClassifier

In [ ]:
%%time
start_time = time.time()

hyperparam = {'max_depth': np.arange(30,51,2),
              'min_samples_split': np.arange(2, 3, 1)}

tree = DecisionTreeClassifier(random_state=12345,
                                    class_weight='balanced')

model_DT = GridSearchCV(tree, hyperparam, scoring='f1', cv=3)

model_DT.fit(features_train, target_train)


print('F1 дерева решений на тренировочной выборке',model_DT.best_score_)

time_dt = (time.time() - start_time)
time_dt

In [ ]:
%%time
start_time = time.time()

print('F1 дерева решений на тестовой выборке', f1_score(target_test, model_DT.predict(features_test)))

time_dt_pred = (time.time() - start_time)
time_dt_pred

### CatBoostClassifier

In [ ]:
scale_pos_weight = round((len(target_train[target_train == 0]) / 
                          len(target_train[target_train == 1])), 3)
scale_pos_weight

In [96]:
%%time
start_time = time.time()

model_cat = CatBoostClassifier(random_state = 12345, verbose=100, eval_metric='F1',
                              scale_pos_weight=scale_pos_weight)



model_cat.fit(features_train, target_train, eval_set=(features_train, target_train))


time_cb = (time.time() - start_time)
time_cb

Learning rate set to 0.12597
0:	learn: 0.4205739	test: 0.4205739	best: 0.4205739 (0)	total: 5.46s	remaining: 1h 30m 59s
100:	learn: 0.8565752	test: 0.8565752	best: 0.8565752 (100)	total: 7m 26s	remaining: 1h 6m 11s
200:	learn: 0.8930349	test: 0.8930349	best: 0.8930349 (200)	total: 14m 44s	remaining: 58m 37s
300:	learn: 0.9135882	test: 0.9135882	best: 0.9135882 (300)	total: 22m 3s	remaining: 51m 14s
400:	learn: 0.9293899	test: 0.9293899	best: 0.9293899 (400)	total: 29m 21s	remaining: 43m 51s
500:	learn: 0.9385680	test: 0.9385680	best: 0.9385680 (500)	total: 36m 39s	remaining: 36m 30s
600:	learn: 0.9449541	test: 0.9449541	best: 0.9449541 (600)	total: 43m 57s	remaining: 29m 11s
700:	learn: 0.9525564	test: 0.9525564	best: 0.9526334 (699)	total: 51m 16s	remaining: 21m 52s
800:	learn: 0.9642326	test: 0.9642326	best: 0.9642326 (800)	total: 58m 35s	remaining: 14m 33s
900:	learn: 0.9759608	test: 0.9759608	best: 0.9759608 (900)	total: 1h 5m 55s	remaining: 7m 14s
999:	learn: 0.9824302	test: 0.982

4528.329458475113

In [97]:
%%time
start_time = time.time()
print('F1 CatBoost на тестовой выборке', f1_score(target_test, model_cat.predict(features_test)))

time_cb_pred = (time.time() - start_time)
time_cb_pred

F1 CatBoost на тестовой выборке 0.7671391379809084
CPU times: user 2.51 s, sys: 713 ms, total: 3.22 s
Wall time: 3.24 s


3.2419142723083496


<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что ты попробовала несколько разных моделей и подходов!
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> С кросс-валидационными методами нужно быть аккуратне, так как внутри кросс-валидации происходит разбиение выборки на треин и валидацию. Однако, в таком случае векторизатор обучен на всей выборке, а это не совсем корректно. Для избежания такого эффекта можно использовать <a href="https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">пайплайн</a>.
</div>

## Выводы

In [99]:
print('F1 Dummy Classifier на тестовой выборке', f1_score(target_test, dummy_clf.predict(features_test)), time_dm)
print('F1 логистической регрессии на тестовой выборке', f1_score(target_test, model_lr.predict(features_test)), time_lr_pred)
print('F1 дерева решений на тестовой выборке', f1_score(target_test, model_DT.predict(features_test)), time_dt_pred)
print('F1 CatBoostClassifier  на тестовой выборке', f1_score(target_test, model_cat.predict(features_test)), time_cb_pred)

F1 Dummy Classifier на тестовой выборке 0.10680644181099971 0.0494692325592041
F1 логистической регрессии на тестовой выборке 0.7513606613847743 0.04588580131530762
F1 дерева решений на тестовой выборке 0.6144306651634723 0.10628080368041992
F1 CatBoostClassifier  на тестовой выборке 0.7671391379809084 3.2419142723083496


Самая лучшая модель это Catboost, f1- 0.767 но она ее обучение требует очень много времени. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Приятно видеть  вывод в конце проекта! 
</div>

<div class="alert alert-block alert-info">
<b>Совет:</b> Здесь хорошо смотрелась бы информативная таблица
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны